### PR TITLE
Dont signal occurrence if client call not active

### PR DESCRIPTION
**Pull request by ni-sujain**
fix for issue #217 

The issue is that client can run into a situation where an inactive client call(e.g. due to vi aborts) will end up signaling the lv occurrence from async thread for an active client call(occurrence is getting re-used) and "wait for occurrence" unblocks without reading the response from server and we will see no response for the active call.

The fix involves clearing the active client calls list on vi abort and signal occurrence only when the current call is active. **This fix will work for most cases but it still can run into some race conditions.**

Alternatives considered:
- Somehow reset the occurrence on vi abort. Could not find a way to do this. Even if a signal the occurrence in `ClientCleanupProc` it doesnot work.
- Reset occurrence before we run ClientUnaryCall CLFN node. Tried _Wait on occurrence_  primitive with zero timeout,  _Set occurence_ primitve but they also did not work. as even if we reset, the async c++ can signal before we go inside new ClientUnaryCall CLFN.
- Do a wait() on blocking client call but this takes away some async behavior.  

testing:
I tried with the attached example from the issue. It runs fine.

### CODE REVIEW

File: src/grpc_client.cc [hunk 1]
```diff
@@ -192,6 +193,22 @@ int32_t CloseClient(grpc_labview::LabVIEWgRPCClient* client)
     return 0;
 }
 
+// Signal a lv occurence for a client call from async c++ thread
+void SignalOccurenceForClientCall(grpc_labview::ClientCall* clientCall)
```
Comment (shivaprasad-basavaraj) [inline src/grpc_client.cc:197]: Should we name this method `CheckActiveAndSignalOccurence` to signify that we will fire the occurrence only if the client call is active?

**Comment by shivaprasad-basavaraj** [comment]
We should add this scenario as an auto test as well.

File: src/grpc_client.cc [hunk 2]
```diff
@@ -218,6 +235,8 @@ int32_t ClientCleanUpProc(grpc_labview::gRPCid* clientId)
         {
             (*activeClientCall)->Cancel();
         }
+
+        client->ActiveClientCalls.clear();
```
Comment (shivaprasad-basavaraj) [inline src/grpc_client.cc:239]: Is `clear()` atomic? Should we get a lock here as well before we clear the list?

**Review by shivaprasad-basavaraj — APPROVED** [review]

File: src/grpc_client.cc (hunk 2 above)
Comment (ni-sujain) [reply src/grpc_client.cc:239]: yes, this is a scoped lock

**Comment by ni-sujain** [comment]
Note: This PR still need some work. There are test failures for streaming cases

**Comment by ni-sujain** [comment]
See PR https://github.com/ni/grpc-labview/pull/251 for this closing this

**Pull request closed without merging**